src/main/java/cubicchunks/asm/mixin/selectable/client/MixinRenderChunk_OptifineSpecific.java

```diff
@@ -0,0 +1,150 @@
+ *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ *  copies of the Software, and to permit persons to whom the Software is
+ *  furnished to do so, subject to the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be included in
+ *  all copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ *  THE SOFTWARE.
+ */
+package cubicchunks.asm.mixin.selectable.client;
+
+import static cubicchunks.client.RenderConstants.*;
+
+import javax.annotation.ParametersAreNonnullByDefault;
+
+import org.spongepowered.asm.mixin.Final;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Overwrite;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Constant;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.ModifyConstant;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
+
+import cubicchunks.client.CubeProviderClient;
+import cubicchunks.util.Coords;
+import cubicchunks.world.ICubicWorldClient;
+import cubicchunks.world.cube.Cube;
+import mcp.MethodsReturnNonnullByDefault;
+import net.minecraft.client.renderer.chunk.RenderChunk;
+import net.minecraft.util.math.BlockPos;
+/**
+ * Fixes renderEntities crashing when rendering cubes
+ * that are not at existing array index in chunk.getEntityLists(),
+ * <p>
+ * Allows to render cubes outside of 0..256 height range.
+ */
+import net.minecraft.world.World;
+import net.minecraft.world.chunk.Chunk;
+
+@MethodsReturnNonnullByDefault
+@ParametersAreNonnullByDefault
+@Mixin(RenderChunk.class)
+public class MixinRenderChunk_OptifineSpecific {
+
+    @Shadow @Final public World world;
+
+    private Cube[] cubeCache;
+    private Chunk[] chunkCache;
+
+    @ModifyConstant(method = "makeChunkCacheOF", constant = @Constant(intValue = 16))
+    public int onMakingChunkCacheOptifine(int oldValue) {
+        return RENDER_CHUNK_SIZE;
+    }
+
+    @Inject(method = "isChunkRegionEmpty(Lnet/minecraft/util/math/BlockPos;)Z", at = @At(value = "HEAD"), cancellable = true, remap = false)
+    public void isChunkRegionEmpty(BlockPos posIn, CallbackInfoReturnable<Boolean> cif) {
+        ICubicWorldClient cworld = (ICubicWorldClient) world;
+        cif.cancel();
+        boolean isEmpty = true;
+        if (cworld.isCubicWorld()) {
```

Comment:
Why duplicate all of the code? Also `chunk.isEmptyBetween` works with cubic chunks.

Reply:
Thats different code. One check if all chunks are empty and other check if entities are present. If entities or empty chunk detected it will skip further checks and only cache them, so there is no possibility to refactor that piece of code.